libbeat/outputs/elasticsearch/client.go

```diff
@@ -380,80 +399,108 @@ func getPipeline(event *beat.Event, defaultSelector *outil.Selector) (string, er
+			stats.fails += count - i
+			eventsToRetry = append(eventsToRetry, events[i:]...)
+			break
 		}
 
-		if status < 300 {
-			stats.acked++
-			continue // ok value
+		if client.applyItemStatus(events[i], itemStatus, itemMessage, &stats) {
+			eventsToRetry = append(eventsToRetry, events[i])
+			log.Debugf("Bulk item insert failed (i=%v, status=%v): %s", i, itemStatus, itemMessage)
 		}
+	}
 
-		if status == 409 {
-			// 409 is used to indicate an event with same ID already exists if
-			// `create` op_type is used.
-			stats.duplicates++
-			continue // ok
-		}
+	return eventsToRetry, stats
+}
 
-		if status < 500 {
-			if status == http.StatusTooManyRequests {
-				stats.tooMany++
-			} else {
-				// hard failure, apply policy action
-				encodedEvent := data[i].EncodedEvent.(*encodedEvent)
-				if encodedEvent.deadLetter {
-					stats.nonIndexable++
-					client.log.Errorf("Can't deliver to dead letter index event (status=%v). Look at the event log to view the event and cause.", status)
-					client.log.Errorw(fmt.Sprintf("Can't deliver to dead letter index event %#v (status=%v): %s", data[i], status, msg), logp.TypeKey, logp.EventType)
-					// poison pill - this will clog the pipeline if the underlying failure is non transient.
-				} else if client.deadLetterIndex != "" {
-					client.log.Warnf("Cannot index event (status=%v), trying dead letter index. Look at the event log to view the event and cause.", status)
-					client.log.Warnw(fmt.Sprintf("Cannot index event %#v (status=%v): %s, trying dead letter index", data[i], status, msg), logp.TypeKey, logp.EventType)
-					client.setDeadLetter(encodedEvent, status, string(msg))
-
-				} else { // drop
-					stats.nonIndexable++
-					client.log.Warnf("Cannot index event (status=%v): dropping event! Look at the event log to view the event and cause.", status)
-					client.log.Warnw(fmt.Sprintf("Cannot index event %#v (status=%v): %s, dropping event!", data[i], status, msg), logp.TypeKey, logp.EventType)
-					continue
-				}
-			}
+// applyItemStatus processes the ingestion status of one event from a bulk request.
+// Returns true if the item should be retried.
+// In the provided bulkResultStats, applyItemStatus increments exactly one of:
+// acked, duplicates, deadLetter, fails, nonIndexable.
+func (client *Client) applyItemStatus(
+	event publisher.Event,
+	itemStatus int,
+	itemMessage []byte,
+	stats *bulkResultStats,
+) bool {
+	encodedEvent := event.EncodedEvent.(*encodedEvent)
+	if itemStatus < 300 {
+		if encodedEvent.deadLetter {
+			// This was ingested into the dead letter index, not the original target
+			stats.deadLetter++
+		} else {
+			stats.acked++
 		}
+		return false // ok value
+	}
 
-		client.log.Debugf("Bulk item insert failed (i=%v, status=%v): %s", i, status, msg)
-		stats.fails++
-		failed = append(failed, data[i])
+	if itemStatus == 409 {
+		// 409 is used to indicate an event with same ID already exists if
+		// `create` op_type is used.
```

Comment:
```suggestion
		// 409 is used to indicate an event with same ID already exists if
		// `create` op_type is used.
```

You now also get a 409 if the TSDB dimensions are the same for two events, they form an implicit ID. This is probably the more common reason to hit this now.

Reply:
Added